content/doc/developer/internationalization/prepare-a-java-source-file.adoc

```diff
@@ -0,0 +1,118 @@
+    // ^Can also use req.bindJSON(this, formData);
+    //  (easier when there are many fields; need set* methods for this, like setUseFrench)
+    save();
+    return super.configure(req,formData);
+}
+
+----
+<1> This string should be internationalized, as it is used as a form validation response.
+<2> This string should also be internationalized, as it is another form validation response.
+<3> This is the display name of the build step on the UI, and should also be internationalized.
+<4> This is an internal identifier for a form configuration option and must not be internationalized.
+
+== Step 2: Prepare the +Messages.properties+ file
+
+
+////
+// TODO better documentation for tool support
+////
+
+[NOTE]
+====
+The link:https://wiki.jenkins-ci.org/display/JENKINS/Internationalization#Internationalization-StaplerpluginforIntelliJIDEA[IntelliJ IDEA Plugin for Stapler] has features that make this and the following steps easier.
+====
+
+The localized strings for the default language (typically English), unless they are reused within the same component, are typically added to a +Messages.properties+ file in the same package as the class that uses it.
+
+[IMPORTANT]
+====
+Do not reference other components (e.g. core's) +Messages+ classes -- they typically do not consider their localized messages stable API, so your plugin might break if they change their messages.
+====
+
+Using the default Maven project layout, the +HelloWorldBuilder+ class stored in the source file +src/main/java/org/jenkinsci/plugins/example/HelloWorldBuilder.java+ would use the localized strings in the resources file +src/main/resources/org/jenkinsci/plugins/example/Messages.properties+ (basically, in the same package). Create that file if it doesn't exist.
+
+Now, let's add entries for the localizable strings identified above to that file:
+
+[source]
+.Messages.properties
+----
+HelloWorldBuilder.NoName=Please set a name // <1>
+HelloWorldBuilder.ShortName=\// <2>
+  Isn''t the name too short? // <3>
+# Unicode variant of the line above: // <4>
+# Isn\u2019t the name too short? // <5>
+HelloWorldBuilder.DisplayName=Say hello world
+----
+<1> This sets the value for the key +HelloWorldBuilder.NoName+. While the key can be freely chosen, it's typically a good idea to indicate where it's used, as multiple files may share the same +Messages.properties+ file. This is also the format generated by e.g. the Stapler Plugin for IntelliJ IDEA.
+<2> Line breaks, mostly useful with very long values, e.g. instructions. can be escaped with a back slash. Make sure to not add whitespace after it.
+<3> In properties files, single quotes need to be escaped with another single quote. Alternatively, a Unicode apostrophe could be used here as well.
+<4> Lines starting with the +#+ character are comments.
+<5> The file encoding for .properties files is ISO-8859-1 (Latin 1), so characters not in that set need to be specified using their Unicode code point, e.g. +\u2019+ for the apostrophe.
```

Comment:
BTW the NetBeans editor transparently encodes/decodes Unicode escapes for `*.properties`.